File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
from sys import stdin
import sys
sys.setrecursionlimit(10 ** 5)

n, m = map(int, stdin.readline().split())
graph = {}
visited = [0 for _ in range(n+1)]
answer = [0]

for i in range(1, n+1):
    graph[i] = []
for _ in range(m):
    u, v = map(int, stdin.readline().split())
    graph[u].append(v)
    graph[v].append(u)


def dfs(curr):
    if visited[curr]==0:
        answer[0] += 1
        visited[curr]=1

    for nxt in graph[curr]:
        if visited[nxt]==1: #다음 노드를 가봤다면 다시 갈 필요 없음
            continue
        else: #다음 노드를 안 가봤다면 연결되었으니 가봤다고 하고 dfs 실행
            visited[nxt]=1
            dfs(nxt)

for g in graph:
    if visited[g]==0:
        dfs(g)
print(answer[0])
    

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.util.ArrayList;
import java.util.Scanner;

public class Main {
    static boolean visited[];
    static ArrayList<Integer>[] A;

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt(); //정점의 개수
        int M = sc.nextInt(); //간선의 개수
        visited = new boolean[N+1];
        A = new ArrayList[N+1];
        for(int i=1; i<N+1; i++){
            A[i] = new ArrayList<Integer>();
        }
        for(int i=0; i<M; i++){
            int u = sc.nextInt();
            int v = sc.nextInt();

            A[u].add(v);
            A[v].add(u);
        }

        int count = 0;
        for(int i=1; i<N+1; i++){
            if(!visited[i]){
                count++;
                DFS(i);
            }
        }
        System.out.println(count);
    }

    private static void DFS(int i){
        if(visited[i]) {
            return;
        }
        visited[i] = true;
        for(int nxt: A[i]){
            if(!visited[nxt]){
                DFS(nxt);
            }
        }
    }
}
